// smartride-frontend/src/components/ui/RideTrackingModal.jsx
import { createPortal } from 'react-dom';
import { useEffect, useRef, useState } from 'react';
import { format } from 'date-fns';
import CancelRideReasonDialog from './CancelRideReasonDialog';
import TripChatDialog from './TripChatDialog';
import RoutePreviewMap from './RoutePreviewMap';
import { rideService } from '../../services/rideService';
import { clockIcon, closeIcon, motorbikeIcon, originIcon, phoneIcon, pinIcon, starIcon, userIcon } from '../../assets/icons';
import { classNames } from '../../utils/classNames';

const SEARCH_TRACKING_STEPS = [
  {
    id: 'created',
    title: 'Đã tạo cuốc xe',
    description: 'Yêu cầu vừa được tiếp nhận và chuyển sang bước ghép tài xế.',
  },
  {
    id: 'matching',
    title: 'Đang tìm tài xế gần đó',
    description: 'Hệ thống ưu tiên các tài xế quanh điểm đón để tạo cuốc nhanh nhất.',
  },
  {
    id: 'confirming',
    title: 'Chờ tài xế xác nhận',
    description: 'Khi có tài xế phù hợp nhận chuyến, cuốc xe sẽ được kích hoạt.',
  },
];

const ASSIGNED_TRACKING_STEPS = [
  {
    id: 'accepted',
    title: 'Tài xế đã nhận đơn',
    description: 'Đơn đã có tài xế xác nhận và hệ thống đang theo dõi lộ trình đến điểm đón.',
  },
  {
    id: 'heading-pickup',
    title: 'Đang đến điểm đón',
    description: 'Tài xế đang di chuyển đến vị trí của bạn trước khi bắt đầu chuyến.',
  },
  {
    id: 'picked-up',
    title: 'Đã đón khách',
    description: 'Khách đã lên xe và chuyến đang được thực hiện.',
  },
];

function formatBookingTime(timestamp) {
  const parsedDate = timestamp ? new Date(timestamp) : null;

  if (!parsedDate || Number.isNaN(parsedDate.getTime())) {
    return '';
  }

  return format(parsedDate, 'HH:mm dd/MM/yyyy');
}

function formatKilometers(distanceKm) {
  const normalizedDistance = Number(distanceKm);

  if (!Number.isFinite(normalizedDistance)) {
    return '';
  }

  return `${normalizedDistance.toFixed(1)} km`;
}

function formatMinutes(minutes) {
  const normalizedMinutes = Number(minutes);

  if (!Number.isFinite(normalizedMinutes)) {
    return '';
  }

  return `${normalizedMinutes} phút`;
}

function formatLicensePlate(value) {
  const normalizedValue = normalizeText(value).toUpperCase();

  if (!normalizedValue) {
    return '';
  }

  return normalizedValue.replace(/\s*-\s*/g, ' - ');
}

function normalizeText(value) {
  return String(value ?? '')
    .trim()
    .replace(/\s+/g, ' ');
}

function extractTripHistoryItems(response) {
  if (Array.isArray(response?.items)) {
    return response.items;
  }

  if (Array.isArray(response?.history)) {
    return response.history;
  }

  if (Array.isArray(response?.data?.items)) {
    return response.data.items;
  }

  if (Array.isArray(response?.data)) {
    return response.data;
  }

  return [];
}

function normalizeTripStatusToken(value) {
  return normalizeText(value)
    .toLowerCase()
    .replace(/[\s_-]+/g, '');
}

function getDriverInitials(name) {
  const parts = normalizeText(name)
    .split(' ')
    .filter(Boolean);

  if (parts.length === 0) {
    return 'TX';
  }

  if (parts.length === 1) {
    return parts[0].slice(0, 2).toUpperCase();
  }

  return `${parts[0].slice(0, 1)}${parts[parts.length - 1].slice(0, 1)}`.toUpperCase();
}

function CopyGlyph() {
  return (
    <svg className="booking-tracking-modal__glyph-icon" viewBox="0 0 16 16" aria-hidden="true" focusable="false">
      <rect x="6" y="5" width="7" height="8" rx="1.1" fill="none" stroke="currentColor" strokeWidth="1.25" />
      <rect x="3" y="3" width="7" height="8" rx="1.1" fill="none" stroke="currentColor" strokeWidth="1.25" />
    </svg>
  );
}

function ChatGlyph() {
  return (
    <svg className="booking-tracking-modal__glyph-icon" viewBox="0 0 16 16" aria-hidden="true" focusable="false">
      <path
        d="M3 3.5h10a1 1 0 0 1 1 1v5a1 1 0 0 1-1 1H8.2L5.3 13v-2.5H3a1 1 0 0 1-1-1v-5a1 1 0 0 1 1-1Z"
        fill="none"
        stroke="currentColor"
        strokeWidth="1.2"
        strokeLinejoin="round"
      />
    </svg>
  );
}

function DriverPlateGlyph() {
  return (
    <svg className="booking-tracking-modal__plate-icon" viewBox="0 0 24 16" aria-hidden="true" focusable="false">
      <rect x="1.2" y="1.4" width="21.6" height="13.2" rx="2.4" fill="none" stroke="currentColor" strokeWidth="1.3" />
      <path d="M4.1 5.2h3.2" fill="none" stroke="currentColor" strokeWidth="1.2" strokeLinecap="round" />
      <path d="M11 5.2h8.8" fill="none" stroke="currentColor" strokeWidth="1.2" strokeLinecap="round" />
      <path d="M4.1 8.6h15.7" fill="none" stroke="currentColor" strokeWidth="1.2" strokeLinecap="round" />
      <path d="M4.1 11.4h9.8" fill="none" stroke="currentColor" strokeWidth="1.2" strokeLinecap="round" />
    </svg>
  );
}

function isDriverAssignedTripStatus(tripStatusToken) {
  return [
    'danhanchuyen',
    'accepted',
    'dangden',
    'headingpickup',
    'dadon',
    'pickedup',
    'hoanthanh',
    'completed',
  ].includes(tripStatusToken);
}

function isDriverPrePickupTripStatus(tripStatusToken) {
  return ['danhanchuyen', 'accepted', 'dangden', 'headingpickup'].includes(tripStatusToken);
}

function isDriverOnTripTripStatus(tripStatusToken) {
  return ['dadon', 'pickedup', 'hoanthanh', 'completed'].includes(tripStatusToken);
}

function getAssignedStepIndex(tripStatusToken) {
  if (tripStatusToken === 'dadon' || tripStatusToken === 'pickedup' || tripStatusToken === 'hoanthanh' || tripStatusToken === 'completed') {
    return 2;
  }

  if (tripStatusToken === 'dangden' || tripStatusToken === 'headingpickup') {
    return 1;
  }

  return 0;
}

function getTripStatusCopy(tripStatusToken, driverName = '') {
  const normalizedDriverName = normalizeText(driverName);

  if (tripStatusToken === 'danhanchuyen' || tripStatusToken === 'accepted') {
    return {
      title: 'Tài xế đã nhận đơn',
      description: normalizedDriverName
        ? `${normalizedDriverName} đang trên đường đến điểm đón.`
        : 'Tài xế đang trên đường đến điểm đón.',
      badge: 'Đã nhận chuyến',
      progress: 58,
    };
  }

  if (tripStatusToken === 'dangden' || tripStatusToken === 'headingpickup') {
    return {
      title: 'Tài xế đang đến điểm đón',
      description: normalizedDriverName
        ? `${normalizedDriverName} đang di chuyển đến vị trí của bạn.`
        : 'Tài xế đang di chuyển đến vị trí của bạn.',
      badge: 'Đang đến điểm đón',
      progress: 74,
    };
  }

  if (tripStatusToken === 'dadon' || tripStatusToken === 'pickedup') {
    return {
      title: 'Tài xế đã đón khách',
      description: normalizedDriverName
        ? `${normalizedDriverName} đã đón bạn và chuyến đang được thực hiện.`
        : 'Chuyến đang được thực hiện.',
      badge: 'Đang di chuyển',
      progress: 90,
    };
  }

  if (tripStatusToken === 'hoanthanh' || tripStatusToken === 'completed') {
    return {
      title: 'Chuyến đã hoàn thành',
      description: 'Cảm ơn bạn đã sử dụng SmartRide.',
      badge: 'Hoàn tất',
      progress: 100,
    };
  }

  if (tripStatusToken === 'dahuy' || tripStatusToken === 'cancelled') {
    return {
      title: 'Chuyến đã hủy',
      description: 'Yêu cầu chuyến không còn hiệu lực.',
      badge: 'Đã hủy',
      progress: 0,
    };
  }

  return {
    title: 'Đang tìm tài xế gần đó',
    description: 'Hệ thống đang ghép chuyến với tài xế phù hợp quanh điểm đón để tạo cuốc xe nhanh nhất.',
    badge: 'Đang ghép chuyến',
    progress: 25,
  };
}

export default function RideTrackingModal({ open = false, booking = null, onClose, onMinimize, onCancel, onNotify }) {
  const [trackingStepIndex, setTrackingStepIndex] = useState(0);
  const [liveBooking, setLiveBooking] = useState(booking);
  const [cancelDialogOpen, setCancelDialogOpen] = useState(false);
  const [chatDialogOpen, setChatDialogOpen] = useState(false);
  const cancelRequestInFlightRef = useRef(false);

  useEffect(() => {
    if (!open) {
      setCancelDialogOpen(false);
      setChatDialogOpen(false);
      setLiveBooking(booking);
      return undefined;
    }

    setLiveBooking(booking);
    setTrackingStepIndex(0);
    setChatDialogOpen(false);

    const firstTimerId = window.setTimeout(() => {
      setTrackingStepIndex(1);
    }, 2200);

    const secondTimerId = window.setTimeout(() => {
      setTrackingStepIndex(2);
    }, 4400);

    return () => {
      window.clearTimeout(firstTimerId);
      window.clearTimeout(secondTimerId);
    };
  }, [booking?.bookingCode, open]);

  useEffect(() => {
    if (!open) {
      return undefined;
    }

    const bookingCode = normalizeText(booking?.bookingCode ?? '');
    const accountId = normalizeText(booking?.customerAccountId ?? booking?.accountId ?? '');

    if (!bookingCode || !accountId) {
      return undefined;
    }

    let isMounted = true;

    const refreshLiveBooking = async () => {
      try {
        const response = await rideService.getTripHistory({
          accountId,
          roleCode: 'Q2',
          limit: 24,
        });

        const matchedBooking = extractTripHistoryItems(response).find((item) => {
          const itemBookingCode = normalizeText(item?.bookingCode ?? item?.id ?? '');
          return itemBookingCode && itemBookingCode === bookingCode;
        });

        if (isMounted && matchedBooking) {
          setLiveBooking((current) => ({
            ...current,
            ...matchedBooking,
          }));
        }
      } catch {
        // Keep the optimistic snapshot if the history refresh fails.
      }
    };

    void refreshLiveBooking();

    const intervalId = window.setInterval(() => {
      void refreshLiveBooking();
    }, 5000);

    return () => {
      isMounted = false;
      window.clearInterval(intervalId);
    };
  }, [booking?.accountId, booking?.bookingCode, booking?.customerAccountId, open]);

  useEffect(() => {
    if (!open) {
      return undefined;
    }

    const handleKeyDown = (event) => {
      if (event.key === 'Escape') {
        if (chatDialogOpen) {
          setChatDialogOpen(false);
          return;
        }

        onClose?.();
      }
    };

    document.addEventListener('keydown', handleKeyDown);

    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [chatDialogOpen, onClose, open]);

  const bookingTimeLabel = formatBookingTime(booking?.createdAt);
  const currentBooking = liveBooking ?? booking;
  const pickupLabel = String(currentBooking?.pickup?.label ?? '').trim() || 'Điểm đón';
  const destinationLabel = String(currentBooking?.destination?.label ?? '').trim() || 'Điểm đến';
  const bookingCode = String(currentBooking?.bookingCode ?? '').trim();
  const rideTitle = String(currentBooking?.rideTitle ?? currentBooking?.vehicleLabel ?? '').trim() || 'Chuyến xe';
  const paymentSummary = String(currentBooking?.paymentSummary ?? currentBooking?.paymentMethodLabel ?? '').trim();
  const paymentStatus = String(currentBooking?.paymentStatusLabel ?? '').trim();
  const priceLabel = String(currentBooking?.priceFormatted ?? '').trim();
  const routeDistanceLabel = formatKilometers(currentBooking?.routeDistanceKm);
  const etaLabel = formatMinutes(currentBooking?.etaMinutes);
  const tripStatusToken = normalizeTripStatusToken(currentBooking?.tripStatus ?? currentBooking?.status ?? '');
  const liveStatus = getTripStatusCopy(tripStatusToken, currentBooking?.driverDisplayName ?? currentBooking?.driverName ?? '');
  const driverName = String(currentBooking?.driverDisplayName ?? currentBooking?.driverName ?? '').trim();
  const driverPhone = String(currentBooking?.driverPhone ?? '').trim();
  const driverVehicleLabel = String(currentBooking?.driverVehicleLabel ?? currentBooking?.vehicleLabel ?? currentBooking?.rideTitle ?? '').trim();
  const driverLicensePlateLabel = formatLicensePlate(currentBooking?.driverLicensePlate ?? currentBooking?.driverVehicleLicensePlate ?? '');
  const driverRatingValue = Number(currentBooking?.driverRating ?? currentBooking?.rating ?? 4) || 4;
  const driverRatingStars = Math.max(0, Math.min(5, Math.round(driverRatingValue)));
  const driverInitials = getDriverInitials(driverName);
  const isPrePickupAssignedState = isDriverPrePickupTripStatus(tripStatusToken);
  const isOnTripState = isDriverOnTripTripStatus(tripStatusToken);
  const isAssignedState = isPrePickupAssignedState || isOnTripState;
  const timelineSteps = isAssignedState ? ASSIGNED_TRACKING_STEPS : SEARCH_TRACKING_STEPS;
  const timelineStepIndex = isAssignedState ? getAssignedStepIndex(tripStatusToken) : trackingStepIndex;
  const progressValue = liveStatus.progress;
  const activeStep = timelineSteps[timelineStepIndex] ?? timelineSteps[timelineSteps.length - 1];
  const canCallDriver = Boolean(normalizeText(driverPhone));
  const driverVehicleDisplayLabel = driverVehicleLabel || 'RiBike';

  const copyTextToClipboard = async (textToCopy) => {
    const normalizedText = normalizeText(textToCopy);

    if (!normalizedText) {
      return false;
    }

    try {
      if (navigator.clipboard?.writeText && window.isSecureContext) {
        await navigator.clipboard.writeText(normalizedText);
        return true;
      }

      const temporaryTextArea = document.createElement('textarea');
      temporaryTextArea.value = normalizedText;
      temporaryTextArea.setAttribute('readonly', 'true');
      temporaryTextArea.style.position = 'fixed';
      temporaryTextArea.style.opacity = '0';
      temporaryTextArea.style.pointerEvents = 'none';
      document.body.appendChild(temporaryTextArea);
      temporaryTextArea.focus();
      temporaryTextArea.select();

      const copied = document.execCommand('copy');
      document.body.removeChild(temporaryTextArea);
      return copied;
    } catch {
      return false;
    }
  };

  const handleChatDriver = () => {
    if (!driverPhone) {
      onNotify?.('Số điện thoại tài xế đang được cập nhật.', 'error', 2200);
      return;
    }

    setCancelDialogOpen(false);
    setChatDialogOpen(true);
  };

  const handleCancel = () => {
    if (!booking || cancelRequestInFlightRef.current) {
      return;
    }

    setChatDialogOpen(false);
    setCancelDialogOpen(true);
  };

  const handleCancelDialogClose = () => {
    if (cancelRequestInFlightRef.current) {
      return;
    }

    setCancelDialogOpen(false);
  };

  const handleCancelDialogConfirm = async (cancelDetails = null) => {
    if (cancelRequestInFlightRef.current) {
      return;
    }

    cancelRequestInFlightRef.current = true;
    setCancelDialogOpen(false);

    try {
      await onCancel?.(cancelDetails);
    } catch (error) {
      onNotify?.(error?.message || 'Không thể hủy chuyến.', 'error', 2200);
    } finally {
      cancelRequestInFlightRef.current = false;
    }
  };

  const handleCloseChatDialog = () => {
    setChatDialogOpen(false);
  };

  const handleBackdropClick = () => {
    setCancelDialogOpen(false);
    setChatDialogOpen(false);

    if (onMinimize) {
      onMinimize();
      return;
    }

    onClose?.();
  };

  const handleCopyText = async (textToCopy, successMessage) => {
    const copied = await copyTextToClipboard(textToCopy);

    if (copied) {
      onNotify?.(successMessage, 'success', 1800);
      return;
    }

    onNotify?.('Không thể sao chép nội dung này trên trình duyệt hiện tại.', 'error', 2200);
  };

  const handleCallDriver = () => {
    if (!canCallDriver) {
      return;
    }

    window.location.href = `tel:${normalizeText(driverPhone).replace(/\s+/g, '')}`;
  };

  if (!open || !booking) {
    return null;
  }

  if (isPrePickupAssignedState) {
    return createPortal(
      <div className={classNames('booking-tracking-modal', 'booking-tracking-modal--assigned-preview')} role="dialog" aria-modal="true" aria-label="Tài xế đã nhận chuyến">
        <div className="booking-tracking-modal__backdrop" onClick={handleBackdropClick} aria-hidden="true" />

        <section className="booking-tracking-modal__window booking-tracking-modal__window--preview">
          <button className="booking-tracking-modal__close" type="button" onClick={() => onClose?.()} aria-label="Đóng">
            <img className="booking-tracking-modal__close-icon" src={closeIcon} alt="" aria-hidden="true" />
          </button>

          <div className="booking-tracking-modal__assigned-preview">
            <div className="booking-tracking-modal__assigned-preview-grid">
              <div className="booking-tracking-modal__assigned-preview-left">
                <div className="booking-tracking-modal__assigned-card booking-tracking-modal__assigned-card--route">
                  <div className="booking-tracking-modal__route-row booking-tracking-modal__route-row--preview">
                    <span className="booking-tracking-modal__route-icon booking-tracking-modal__route-icon--pickup booking-tracking-modal__route-icon--preview">
                      <img className="booking-tracking-modal__route-icon-img" src={originIcon} alt="" aria-hidden="true" />
                    </span>

                    <div className="booking-tracking-modal__route-copy booking-tracking-modal__route-copy--preview">
                      <span>Điểm đón</span>
                      <strong>{pickupLabel}</strong>
                    </div>
                  </div>

                  <div className="booking-tracking-modal__route-divider booking-tracking-modal__route-divider--preview" />

                  <div className="booking-tracking-modal__route-row booking-tracking-modal__route-row--preview">
                    <span className="booking-tracking-modal__route-icon booking-tracking-modal__route-icon--destination booking-tracking-modal__route-icon--preview">
                      <img className="booking-tracking-modal__route-icon-img" src={pinIcon} alt="" aria-hidden="true" />
                    </span>

                    <div className="booking-tracking-modal__route-copy booking-tracking-modal__route-copy--preview">
                      <span>Điểm đến</span>
                      <strong>{destinationLabel}</strong>
                    </div>
                  </div>
                </div>

                <div className="booking-tracking-modal__assigned-card booking-tracking-modal__assigned-card--trip">
                  <div className="booking-tracking-modal__trip-head booking-tracking-modal__trip-head--preview">
                    <div>
                      <span className="booking-tracking-modal__trip-kicker booking-tracking-modal__trip-kicker--preview">{rideTitle}</span>
                    </div>

                    <button className="booking-tracking-modal__booking-code booking-tracking-modal__booking-code--preview" type="button" onClick={() => void handleCopyText(bookingCode, 'Đã sao chép mã chuyến.') } aria-label="Sao chép mã chuyến">
                      <span>{bookingCode || '--'}</span>
                      <CopyGlyph />
                    </button>
                  </div>

                  <div className="booking-tracking-modal__trip-preview-grid">
                    <div className="booking-tracking-modal__trip-preview-row">
                      <span className="booking-tracking-modal__trip-preview-label">Thời gian</span>
                      <strong className="booking-tracking-modal__trip-preview-value">{bookingTimeLabel || '--'}</strong>
                    </div>

                    <div className="booking-tracking-modal__trip-preview-row booking-tracking-modal__trip-preview-row--payment">
                      <span className="booking-tracking-modal__payment-chip">{paymentSummary || 'Tiền mặt'}</span>
                      <strong className="booking-tracking-modal__trip-preview-value booking-tracking-modal__trip-preview-value--price">{priceLabel || '--'}</strong>
                    </div>
                  </div>
                </div>
              </div>

              <div className="booking-tracking-modal__assigned-preview-right">
                <div className="booking-tracking-modal__driver-card">
                  <div className="booking-tracking-modal__driver-avatar booking-tracking-modal__driver-avatar--preview" aria-hidden="true">
                    <img className="booking-tracking-modal__driver-avatar-icon" src={userIcon} alt="" aria-hidden="true" />
                  </div>

                  <strong className="booking-tracking-modal__driver-name">{driverName || 'Nguyễn Văn A'}</strong>

                  <div className="booking-tracking-modal__driver-rating" aria-label={`Đánh giá ${driverRatingStars} trên 5 sao`}>
                    {Array.from({ length: 5 }).map((_, index) => (
                      <img
                        key={`${index + 1}`}
                        className={classNames('booking-tracking-modal__driver-rating-star', index >= driverRatingStars && 'is-muted')}
                        src={starIcon}
                        alt=""
                        aria-hidden="true"
                      />
                    ))}
                  </div>

                  <div className="booking-tracking-modal__contact-row">
                    <div className="booking-tracking-modal__phone-block">
                      <img className="booking-tracking-modal__inline-icon" src={phoneIcon} alt="" aria-hidden="true" />
                      <span className="booking-tracking-modal__phone-number">{driverPhone || 'Đang cập nhật'}</span>
                      <button className="booking-tracking-modal__icon-button" type="button" onClick={() => void handleCopyText(driverPhone, 'Đã sao chép số điện thoại tài xế.') } aria-label="Sao chép số điện thoại tài xế">
                        <CopyGlyph />
                      </button>
                    </div>

                    <button className="booking-tracking-modal__chat-button" type="button" onClick={handleChatDriver} disabled={!driverPhone}>
                      <ChatGlyph />
                      <span>Chat</span>
                    </button>
                  </div>

                  <div className="booking-tracking-modal__driver-line">
                    <img className="booking-tracking-modal__inline-icon booking-tracking-modal__inline-icon--vehicle" src={motorbikeIcon} alt="" aria-hidden="true" />
                    <span>{driverVehicleDisplayLabel}</span>
                  </div>

                  <div className="booking-tracking-modal__driver-line">
                    <DriverPlateGlyph />
                    <span>{driverLicensePlateLabel || 'Đang cập nhật'}</span>
                  </div>
                </div>

                <footer className="booking-tracking-modal__preview-actions">
                  <button className="booking-tracking-modal__cancel booking-tracking-modal__cancel--preview" type="button" onClick={handleCancel}>
                    Hủy chuyến
                  </button>

                  <button className="booking-tracking-modal__location-button" type="button">
                    Vị trí tài xế
                  </button>
                </footer>

                <p className="booking-tracking-modal__preview-note">Hủy chuyến sau khi tài xế nhận có thể phát sinh chi phí</p>
              </div>
            </div>
          </div>
        </section>

        <CancelRideReasonDialog
          open={cancelDialogOpen}
          onCancel={handleCancelDialogClose}
          onConfirm={handleCancelDialogConfirm}
        />

        <TripChatDialog
          open={chatDialogOpen}
          bookingCode={bookingCode}
          accountId={String(currentBooking?.customerAccountId ?? currentBooking?.accountId ?? '').trim()}
          roleCode="Q2"
          dialogTitle="Liên hệ tài xế"
          dialogSubtitle={`${driverVehicleDisplayLabel}${driverLicensePlateLabel ? ` · ${driverLicensePlateLabel}` : ''}`}
          statusLabel="Đang tới điểm đón"
          statusValue={driverPhone || 'Đang cập nhật'}
          contactName={driverName || 'Tài xế'}
          contactPhone={driverPhone}
          quickReplies={[
            'Dạ anh tới gần chưa ạ?',
            'Em đang đứng ở cổng chính ạ.',
            'Anh tới nơi gọi em giúp nhé.',
          ]}
          onClose={handleCloseChatDialog}
          onNotify={onNotify}
        />
      </div>,
      document.body,
    );
  }

  return createPortal(
    <div className={classNames('booking-tracking-modal', isAssignedState && 'booking-tracking-modal--assigned')} role="dialog" aria-modal="true" aria-label={isOnTripState ? 'Tài xế đã đón khách' : isAssignedState ? 'Tài xế đã nhận chuyến' : 'Trạng thái tìm tài xế'}>
      <div className="booking-tracking-modal__backdrop" onClick={handleBackdropClick} aria-hidden="true" />

      <section className="booking-tracking-modal__window">
        <button className="booking-tracking-modal__close" type="button" onClick={() => onClose?.()} aria-label="Đóng">
          <img className="booking-tracking-modal__close-icon" src={closeIcon} alt="" aria-hidden="true" />
        </button>

        <header className="booking-tracking-modal__header">
          <p className="booking-tracking-modal__eyebrow">{isAssignedState ? 'Tài xế đã nhận chuyến' : 'Đặt xe thành công'}</p>
          <h3>{liveStatus.title}</h3>
          <p className="booking-tracking-modal__description">{liveStatus.description}</p>
          {isAssignedState ? (
            <div className="booking-tracking-modal__status-chip-row" aria-label="Trạng thái chuyến xe">
              <span className="booking-tracking-modal__status-chip booking-tracking-modal__status-chip--active">Đã có tài xế</span>
              <span className="booking-tracking-modal__status-chip">Chưa đón khách</span>
              <span className="booking-tracking-modal__status-chip booking-tracking-modal__status-chip--tone">Theo dõi lộ trình</span>
            </div>
          ) : null}
        </header>

        <div className="booking-tracking-modal__body">
          <div className="booking-tracking-modal__left">
            <div className="booking-tracking-modal__route-card">
              <div className="booking-tracking-modal__route-row">
                <span className="booking-tracking-modal__route-icon booking-tracking-modal__route-icon--pickup">
                  <img className="booking-tracking-modal__route-icon-img" src={originIcon} alt="" aria-hidden="true" />
                </span>

                <div className="booking-tracking-modal__route-copy">
                  <span>Điểm đón</span>
                  <strong>{pickupLabel}</strong>
                </div>
              </div>

              <div className="booking-tracking-modal__route-divider" />

              <div className="booking-tracking-modal__route-row">
                <span className="booking-tracking-modal__route-icon booking-tracking-modal__route-icon--destination">
                  <img className="booking-tracking-modal__route-icon-img" src={pinIcon} alt="" aria-hidden="true" />
                </span>

                <div className="booking-tracking-modal__route-copy">
                  <span>Điểm đến</span>
                  <strong>{destinationLabel}</strong>
                </div>
              </div>
            </div>

            <div className="booking-tracking-modal__trip-card">
              <div className="booking-tracking-modal__trip-head">
                <div>
                  <span className="booking-tracking-modal__trip-kicker">{booking?.vehicleLabel?.trim() || 'Cuốc xe'}</span>
                  <h4>{rideTitle}</h4>
                </div>

                <div className="booking-tracking-modal__booking-code">
                  <span>Mã chuyến</span>
                  <strong>{bookingCode || '--'}</strong>
                </div>
              </div>

              <div className="booking-tracking-modal__details">
                <div className="booking-tracking-modal__detail-item">
                  <span className="booking-tracking-modal__detail-label">
                    <img className="booking-tracking-modal__detail-icon" src={clockIcon} alt="" aria-hidden="true" />
                    Thời gian
                  </span>
                  <strong>{bookingTimeLabel || '--'}</strong>
                </div>

                <div className="booking-tracking-modal__detail-item">
                  <span className="booking-tracking-modal__detail-label">Thanh toán</span>
                  <strong>{paymentSummary || '--'}</strong>
                </div>

                <div className="booking-tracking-modal__detail-item">
                  <span className="booking-tracking-modal__detail-label">Trạng thái</span>
                  <strong>{paymentStatus || '--'}</strong>
                </div>

                <div className="booking-tracking-modal__detail-item">
                  <span className="booking-tracking-modal__detail-label">Giá cước</span>
                  <strong>{priceLabel || '--'}</strong>
                </div>

                <div className="booking-tracking-modal__detail-item">
                  <span className="booking-tracking-modal__detail-label">Quãng đường</span>
                  <strong>{routeDistanceLabel || '--'}</strong>
                </div>

                <div className="booking-tracking-modal__detail-item">
                  <span className="booking-tracking-modal__detail-label">Dự kiến</span>
                  <strong>{etaLabel || '--'}</strong>
                </div>
              </div>
            </div>
          </div>

          <div className="booking-tracking-modal__right">
            {isAssignedState ? (
              <div className="booking-tracking-modal__driver-spotlight">
                <div className="booking-tracking-modal__driver-avatar" aria-hidden="true">
                  {driverInitials}
                </div>

                <div className="booking-tracking-modal__driver-copy">
                  <span>Đã nhận chuyến</span>
                  <strong>{driverName || 'Tài xế đang trên đường'}</strong>
                  <p>{driverVehicleLabel || 'Đang cập nhật loại xe'}{driverPhone ? ` · ${driverPhone}` : ''}</p>
                </div>

                <button
                  className="booking-tracking-modal__call-driver"
                  type="button"
                  onClick={handleCallDriver}
                  disabled={!canCallDriver}
                >
                  <img className="booking-tracking-modal__call-driver-icon" src={phoneIcon} alt="" aria-hidden="true" />
                  {canCallDriver ? 'Gọi tài xế' : 'Đang cập nhật'}
                </button>
              </div>
            ) : null}

            <div className="booking-tracking-modal__map-shell">
              <RoutePreviewMap
                className="booking-tracking-modal__map"
                pickupPosition={currentBooking?.pickup?.position}
                destinationPosition={currentBooking?.destination?.position}
                routeGeometry={currentBooking?.routeGeometry}
                routeProvider={currentBooking?.routeProvider}
                showExpandButton={false}
                showProviderLabel={false}
              />

              <div className="booking-tracking-modal__map-badge">{liveStatus.badge}</div>
            </div>

            <div className={classNames('booking-tracking-modal__status-card', isAssignedState && 'booking-tracking-modal__status-card--assigned')}>
              <div className="booking-tracking-modal__status-head">
                <span className="booking-tracking-modal__status-dot" aria-hidden="true" />
                <strong>{liveStatus.title}</strong>
              </div>

              <p>{liveStatus.description}</p>

              <div className="booking-tracking-modal__status-highlight" aria-label="Tiến trình hiện tại">
                <span>Bước hiện tại</span>
                <strong>{activeStep.title}</strong>
                <p>{activeStep.description}</p>
              </div>

              {driverName ? (
                <div className="booking-tracking-modal__details">
                  <div className="booking-tracking-modal__detail-item">
                    <span className="booking-tracking-modal__detail-label">Tài xế</span>
                    <strong>{driverName}</strong>
                  </div>

                  <div className="booking-tracking-modal__detail-item">
                    <span className="booking-tracking-modal__detail-label">Liên hệ</span>
                    <strong>{driverPhone || 'Đang cập nhật'}</strong>
                  </div>
                </div>
              ) : null}

              <div className="booking-tracking-modal__progress" role="progressbar" aria-valuemin={0} aria-valuemax={100} aria-valuenow={progressValue}>
                <span style={{ width: `${progressValue}%` }} />
              </div>

              <div className="booking-tracking-modal__timeline" aria-label={isAssignedState ? 'Tiến trình đón khách' : 'Tiến trình tìm tài xế'}>
                {timelineSteps.map((step, index) => (
                  <div
                    key={step.id}
                    className={classNames(
                      'booking-tracking-modal__timeline-item',
                      index < timelineStepIndex && 'is-complete',
                      index === timelineStepIndex && 'is-active',
                    )}
                  >
                    <span className="booking-tracking-modal__timeline-index">{index + 1}</span>
                    <div>
                      <strong>{step.title}</strong>
                      <p>{step.description}</p>
                    </div>
                  </div>
                ))}
              </div>
            </div>
          </div>
        </div>

        <footer className={classNames('booking-tracking-modal__footer', isAssignedState && 'booking-tracking-modal__footer--assigned')}>
          {isAssignedState ? (
            <>
              <button className="booking-tracking-modal__call-driver booking-tracking-modal__call-driver--footer" type="button" onClick={handleCallDriver} disabled={!canCallDriver}>
                <img className="booking-tracking-modal__call-driver-icon" src={phoneIcon} alt="" aria-hidden="true" />
                {canCallDriver ? 'Gọi tài xế' : 'Đang cập nhật'}
              </button>

              <button className="booking-tracking-modal__cancel" type="button" onClick={handleCancel}>
                Hủy chuyến
              </button>
            </>
          ) : (
            <button className="booking-tracking-modal__cancel" type="button" onClick={handleCancel}>
              Hủy chuyến
            </button>
          )}
        </footer>
      </section>

      <CancelRideReasonDialog
        open={cancelDialogOpen}
        onCancel={handleCancelDialogClose}
        onConfirm={handleCancelDialogConfirm}
      />
    </div>,
    document.body,
  );
}
