// smartride-backend/src/config/env.js
function parseBoolean(value, fallback = false) {
  if (value === undefined || value === null || String(value).trim() === '') {
    return fallback;
  }

  const normalized = String(value).trim().toLowerCase();
  return normalized === '1' || normalized === 'true' || normalized === 'yes' || normalized === 'on';
}

export const env = {
  port: Number(process.env.PORT ?? 4000),
  corsOrigin: process.env.CORS_ORIGIN ?? 'http://localhost:5173',
  rabbitMqUrl: process.env.RABBITMQ_URL ?? '',
  rabbitMqExchangeName: process.env.RABBITMQ_EXCHANGE_NAME ?? 'smartride.ride.events',
  googleMapsServerApiKey: process.env.GOOGLE_MAPS_SERVER_API_KEY ?? '',
  googleAuthClientId: process.env.GOOGLE_AUTH_CLIENT_ID ?? '',
  smtpHost: process.env.SMTP_HOST ?? '',
  smtpPort: Number(process.env.SMTP_PORT ?? 587),
  smtpSecure: parseBoolean(process.env.SMTP_SECURE, false),
  smtpTlsRejectUnauthorized: parseBoolean(process.env.SMTP_TLS_REJECT_UNAUTHORIZED, true),
  smtpUser: process.env.SMTP_USER ?? '',
  smtpPassword: process.env.SMTP_PASSWORD ?? '',
  smtpFromName: process.env.SMTP_FROM_NAME ?? 'SmartRide',
  smtpFromEmail: process.env.SMTP_FROM_EMAIL ?? '',
  signupOtpExpiresMinutes: Number(process.env.SIGNUP_OTP_EXPIRES_MINUTES ?? 10),
  signupOtpResendCooldownSeconds: Number(process.env.SIGNUP_OTP_RESEND_COOLDOWN_SECONDS ?? 60),
  signupOtpMaxVerifyAttempts: Number(process.env.SIGNUP_OTP_MAX_VERIFY_ATTEMPTS ?? 5),
  googlePasswordTokenExpiresMinutes: Number(process.env.GOOGLE_PASSWORD_TOKEN_EXPIRES_MINUTES ?? 10),
  dbHost: process.env.DB_HOST ?? 'localhost',
  dbInstanceName: process.env.DB_INSTANCE_NAME ?? '',
  dbPort: Number(process.env.DB_PORT ?? 1433),
  dbName: process.env.DB_NAME ?? '',
  dbUser: process.env.DB_USER ?? '',
  dbPassword: process.env.DB_PASSWORD ?? '',
  dbEncrypt: parseBoolean(process.env.DB_ENCRYPT, false),
  dbTrustServerCertificate: parseBoolean(process.env.DB_TRUST_SERVER_CERTIFICATE, true),
  dbPoolMax: Number(process.env.DB_POOL_MAX ?? 10),
  dbConnectionTimeoutMs: Number(process.env.DB_CONNECTION_TIMEOUT_MS ?? 15000),
  dbRequestTimeoutMs: Number(process.env.DB_REQUEST_TIMEOUT_MS ?? 15000),
  dbConnectionRetryCooldownMs: Number(process.env.DB_CONNECTION_RETRY_COOLDOWN_MS ?? 30000),
  driverPlatformFeePercent: Number(process.env.DRIVER_PLATFORM_FEE_PERCENT ?? 30),
  geminiApiKey: process.env.GEMINI_API_KEY ?? '',
  geminiModel: process.env.GEMINI_MODEL ?? 'gemini-2.0-flash',
  geminiModelCandidates: process.env.GEMINI_MODEL_CANDIDATES ?? 'gemini-2.0-flash,gemini-2.5-flash',
  geminiTimeoutMs: Number(process.env.GEMINI_TIMEOUT_MS ?? 9000),
  zaloPayAppId: process.env.ZALOPAY_APP_ID ?? '',
  zaloPayKey1: process.env.ZALOPAY_KEY1 ?? '',
  zaloPayKey2: process.env.ZALOPAY_KEY2 ?? '',
  zaloPayCreateOrderUrl: process.env.ZALOPAY_CREATE_ORDER_URL ?? 'https://sb-openapi.zalopay.vn/v2/create',
  zaloPayQueryOrderUrl: process.env.ZALOPAY_QUERY_ORDER_URL ?? 'https://sb-openapi.zalopay.vn/v2/query',
  zaloPayCallbackUrl: process.env.ZALOPAY_CALLBACK_URL ?? '',
  zaloPayRedirectUrl: process.env.ZALOPAY_REDIRECT_URL ?? '',
  momoPartnerCode: process.env.MOMO_PARTNER_CODE ?? '',
  momoAccessKey: process.env.MOMO_ACCESS_KEY ?? '',
  momoSecretKey: process.env.MOMO_SECRET_KEY ?? '',
  momoCreateOrderUrl: process.env.MOMO_CREATE_ORDER_URL ?? 'https://test-payment.momo.vn/v2/gateway/api/create',
  momoQueryOrderUrl: process.env.MOMO_QUERY_ORDER_URL ?? 'https://test-payment.momo.vn/v2/gateway/api/query',
  momoCallbackUrl: process.env.MOMO_CALLBACK_URL ?? '',
  momoRedirectUrl: process.env.MOMO_REDIRECT_URL ?? '',
  momoRequestType: process.env.MOMO_REQUEST_TYPE ?? 'captureWallet',
  momoMockMode: parseBoolean(process.env.MOMO_MOCK_MODE, false),
};
